pkg/hub/managedclusterset/controller.go

```diff
@@ -41,55 +43,69 @@ func NewManagedClusterSetController(
 
 	controllerName := "managed-clusterset-controller"
 	syncCtx := factory.NewSyncContext(controllerName, recorder)
+
+	c := &managedClusterSetController{
+		clusterClient:    clusterClient,
+		clusterLister:    clusterInformer.Lister(),
+		clusterSetLister: clusterSetInformer.Lister(),
+		eventRecorder:    recorder.WithComponentSuffix("managed-cluster-set-controller"),
+		queue:            syncCtx.Queue(),
+	}
 	clusterInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
 		AddFunc: func(obj interface{}) {
-			if clusterSetName := getClusterSet(obj); len(clusterSetName) > 0 {
-				syncCtx.Queue().Add(clusterSetName)
+			cluster, ok := obj.(*v1.ManagedCluster)
+			if !ok {
+				utilruntime.HandleError(fmt.Errorf("error to get object: %v", obj))
 				return
 			}
+			//enqueue all cluster related clustersets
+			c.enqueueClusterClusterSet(cluster)
 		},
 		UpdateFunc: func(oldObj, newObj interface{}) {
-			oldClusterSetName, newClusterSetName := getClusterSet(oldObj), getClusterSet(newObj)
-			if oldClusterSetName == newClusterSetName {
+			// only need handle label update
+			oldCluster, ok := oldObj.(*v1.ManagedCluster)
+			if !ok {
+				utilruntime.HandleError(fmt.Errorf("error to get object: %v", oldObj))
 				return
 			}
-
-			if len(oldClusterSetName) > 0 {
-				syncCtx.Queue().Add(oldClusterSetName)
+			newCluster, ok := newObj.(*v1.ManagedCluster)
+			if !ok {
+				utilruntime.HandleError(fmt.Errorf("error to get object: %v", newObj))
+				return
 			}
-
-			if len(newClusterSetName) > 0 {
-				syncCtx.Queue().Add(newClusterSetName)
+			if reflect.DeepEqual(oldCluster.Labels, newCluster.Labels) {
+				return
 			}
+			c.enqueueUpdateClusterClusterSet(oldCluster, newCluster)
 		},
 		DeleteFunc: func(obj interface{}) {
-			var clusterSetName string
 			switch t := obj.(type) {
-			case *clusterapiv1.ManagedCluster:
-				clusterSetName = getClusterSet(obj)
+			case *v1.ManagedCluster:
+				cluster, ok := obj.(*v1.ManagedCluster)
+				if !ok {
+					utilruntime.HandleError(fmt.Errorf("error to get object: %v", obj))
+					return
+				}
```

Comment:
You may not need those lines. Instead use `t` as `*clusterapiv1.ManagedCluster` directly.